### ISSUE TITLE
Add abstract methods to ECCurve.AbstractFp, AbstractF2m

### ISSUE THREAD

**Issue by stevemit** [bcgit/bc-java#417]
It would be convenient if the classes ECCurve.AbstractFp and ECCurve.AbstractF2m (org.bouncycastle.math.ec) had abstract methods mirroring those methods that are common to all of their subtypes, for example, getQ, getM, isTriangular, etc.  The method isKoblitz is already there.  A workaround is to use reflection.

**Comment by peterdettman** [comment]
Sounds fine, we'd welcome a PR on that, else we'll get to it at some point. If you can share any details about why you're doing it, there may be additional options also.

**Comment by stevemit** [comment]
I’m converting org.bouncycastle.asn1.x9.X9ECParameters to java.security.spec.ECParameterSpec.  The tricky part is getting a java.security.spec.ECField (ECFieldFp or ECFieldF2m) from org.bouncycastle.math.ec.ECCurve.

Maybe this already exists somewhere?

Currently I test whether the curve inherits from AbstractFp or AbstractF2m and then collect the parameters (q or m, k1, k2, k3), using reflection.  I suppose I could instead check if ECCurve.getField() inherits from PrimeField or PolynomialExtensionField and use q = characteristic, m = dimension, and {k1, k2, k3} = polynomial.

  --Steve

**Comment by peterdettman** [comment]
Yes, `getField()` is better for that sort of thing, see one of the `X9ECParameters` constructors for how to analyse the return values.

However, we have utility classes that handle a lot of those sort of conversions already (needed by the BC provider), including what you're after:

```
    package org.bouncycastle.jcajce.provider.asymmetric.util;
    public class EC5Util {
        ...
        public static ECParameterSpec convertToSpec(X9ECParameters domainParameters) { ... }
        ...
        public static ECField convertField(FiniteField field) { ... }
        ...
    }

```

**Comment by stevemit** [comment]
Thanks, I’ll use your implementation!

  --Steve

**Comment by CardContact** [comment]
I'd still vote to add getQ(), as there doesn't seem to be a clean way to extract Q from a public key if curve definitions from org.bouncycastle.math.ec.custom.sec are used (which is the standard case).

We faced that problem recently, when we migrated in OpenSCDP from BC 1.46 to 1.60. In the scripting environment we allow users to extract individual elements of the key, thus we need access to Q as one parameter of the curve.

We now solved that by adding a switch statement that selectively casts one of the Secxxxx class and calls getQ(). That is not the optimal solution.
